internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,390 @@
+	log := logf.FromContext(ctx)
+
+	nodeTemplate := &nyallocatorv1.NodeTemplate{}
+	err = r.Client.Get(ctx, client.ObjectKey{Name: req.Name}, nodeTemplate)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// add finalizer if not exists
+	if nodeTemplate.DeletionTimestamp == nil && nodeTemplate.Finalizers == nil {
+		nodeTemplate.Finalizers = []string{NodeTemplateFinalizer}
+		err = r.Update(ctx, nodeTemplate)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	existingNodes := corev1.NodeList{}
+	err = r.Client.List(ctx, &existingNodes, client.MatchingLabels{NodeTemplateReferenceLabelKey: nodeTemplate.Name})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if nodeTemplate.DeletionTimestamp != nil {
+		for _, node := range existingNodes.Items {
+			// remove the reference label
+			delete(node.Labels, NodeTemplateReferenceLabelKey)
+			log.Info("node template is being deleted, removing reference label",
+				"node", node.Name,
+				"NodeTemplate", nodeTemplate.Name,
+			)
+			err = r.Client.Update(ctx, &node)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// remove finalizer
+		nodeTemplate.Finalizers = nil
+		log.Info("removing finalizer", "NodeTemplate", nodeTemplate.Name, "finalizer", nodeTemplate.Finalizers)
+		err = r.Update(ctx, nodeTemplate)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// remove metrics
+		r.removeMetrics(nodeTemplate)
+		return ctrl.Result{}, nil
+	}
+
+	defer func() {
+		updateStatusErr := r.updateStatus(ctx, nodeTemplate)
+		if updateStatusErr != nil {
+			err = updateStatusErr
+		}
+		// if nodeTemplate.Status.Sufficient is not true, requeue after 1 second
+		if !nodeTemplate.Status.Sufficient {
+			result = ctrl.Result{
+				RequeueAfter: time.Second * 1,
+			}
+		}
+	}()
+
+	// sync existing nodes
+	for _, node := range existingNodes.Items {
+		nodeOriginal := node.DeepCopy()
+		r.reflectNodeConfiguration(&node, nodeTemplate)
+		if !reflect.DeepEqual(*nodeOriginal, node) {
+			log.Info("updating node",
+				"node", node.Name,
+				"NodeTemplate", nodeTemplate.Name,
+			)
+			if nodeTemplate.Spec.DryRun {
+				log.Info("dry run mode, skipping update nodes.\n diff:\n"+calcDiff(nodeOriginal, &node),
+					"node", node.Name,
+					"NodeTemplate", nodeTemplate.Name,
+					"phase", "updating",
+				)
+				continue
+			}
+			err = r.Client.Update(ctx, &node)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// if there are NodeTemplates that is lack of nodes with higher priority, skip ensuring of nodes.
+	nodeTemplates := nyallocatorv1.NodeTemplateList{}
+	err = r.Client.List(ctx, &nodeTemplates)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, nt := range nodeTemplates.Items {
+		if nt.Spec.Priority > nodeTemplate.Spec.Priority && !nt.Status.Sufficient {
+			log.Info("skip ensuring node, because there is a NodeTemplate with higher priority that is not sufficient",
+				"target", nodeTemplate.Name,
```

Comment:
In other logs, `NodeTemplate` is used.
```suggestion
				"NodeTemplate", nodeTemplate.Name,
```